### ISSUE TITLE
MSA validation sometimes shows 'Unexpected error'

### ISSUE THREAD

**Issue by yeban** [wurmlab/genevalidator#75]
This is most likely due to the OS randomly killing MAFFT processes because the system ran out of memory. I think I mentioned somewhere that MAFFT processes can consume unusually high amounts of memory, especially when the number of sequences to align is high (like, 500 blast hits).

This could be a bit of a problem, especially when running gv on desktop computers (like my iMac with 8 GB memory). I don't think it's a critical issue, but should be resolved eventually. In the meantime it might be worth documenting this somewhere - perhaps advising users running into this issue to run gv on a fat machine or to limit the number of BLAST hits to 100 or so and provide the XML to gv with `-x` option.

**Comment by yannickwurm** [comment]
didn't we previously limit to the top 10 or 20 as shown on the graph?
++
yannick

**Comment by monicadragan** [comment]
I am not sure the problem is because of OOM: see below how the head of the output for 30,000 genes ran on Prometheus last month looks like  (and 'Unexpected error' appears among the first queries).

Another possible cause might be in the raw_sequence retrieval step: if it's not possible to retrieve the raw sequence of one hit (from internet or from a provided database) for different reasons (e.g accession numbers don't match), the whole validation fails (that hit might be ignored instead).

****

 No|Score|          Identifier|No_Hits|LengthCluster|LengthRank|Gene_Merge|Duplication|||MA
  1|75|Translation:ENSMUSP00000000001|  183|353 [354, 355]|15% (too short)|0.0|1.0|||99% conserved; 0% extra; 0% missing.|
  2|100|Translation:ENSMUSP00000000003|  188|174 [167, 181]|43%|-0.1|Unexpected error|||Unexpected error|
  4|100|Translation:ENSMUSP00000000012|  189|188 [187, 188]|15% (too long)|0.0|Unexpected error|||Unexpected error|
  3|100|Translation:ENSMUSP00000000010|  188|250 [231, 261]|27%|0.0|1.0|||100% conserved; 0% extra; 0% missing.|
  6|66|Translation:ENSMUSP00000000031|    9|147 [131, 197]|22%|0.0|Unexpected error|||Not enough evidence|
  5|100|Translation:ENSMUSP00000000028|  187|566 [557, 589]|42%|0.0|1.0|||100% conserved; 0% extra; 0% missing.|
  7|100|Translation:ENSMUSP00000000033|  188|180 [164, 225]|24%|0.0|1.0|||99% conserved; 0% extra; 1% missing.|
  8|50|Translation:ENSMUSP00000000034|  188|385 [681, 702]|1% (too short)|0.0|1.0|||89% conserved; 0% extra; 44% missing.|
  9|50|Translation:ENSMUSP00000000040|  142|225 [372, 708]|16% (too short)|0.6|1.0|||100% conserved; 3% extra; 3% missing.|

**Comment by IsmailM** [comment]
@yannickwurm  - MSA validation takes the top 10 hits and creates an alignment from them...

@monicadragan , any chance you could take some of the sequences that previously showed 'unexpected error' and re-run with the latest master (i.e. install from github)...

**Comment by IsmailM** [comment]
Is this still a issue or can we close this issue?

If so, please provide me with the input_file(s) used...

**Comment by monicadragan** [comment]
This was observed when running large scale input (e.g., 30,000 queries) on the grid:
http://swarm.cs.pub.ro/~mdragan/gsoc2013/genevalidator/

**Comment by nldowell** [comment]
I am getting the same error ("Unexpected error") during my use of genevalidator. The program seems to be proceeding fine up to the multiple alignment stage (see below).

Has any solution been proposed? Thanks in advance for any tips.

My use case:
I have assembled a medium-sized genome of a vertebrate. I ran geneID on the repeat masked genome to get a crude annotation. I am now trying to find "good" gene models with genevalidator. I broke the predicted geneID protein sequence FastA files into chunks (~6K - 12K sequences) before BLASTing a local UniProt50 database. 

What I have done:
I then extracted the hit sequences from the blastDB using the genevalidator command and option 'e'. Next I passed the blast-hit file, plus 'raw seqs', plus input seqs to the genevalidotor command:

genevalidator -d /home/snaketeamalpha/Documents/noah/RefSeqData/uniRef50/uniRef50ProtDB -n 40 \
-t ../blastUniRef50/scutB_orthoSetA.geneID_proteins.fasta.uf50hits.txt \
-b /home/snaketeamalpha/lib/ncbi-blast-2.3.0+/bin/ \
-b /usr/bin/ \
-o 'qseqid sseqid sacc slen qstart qend sstart send length qframe pident nident evalue qseq sseq' \
-r ../blastUniRef50/scutB_orthoSetA.geneID_proteins.fasta.uf50hits.txt.raw_seq \
../proteinSeqs/scutB_orthoSetA.geneID_proteins.fasta > scutB_orthoSetA_geneValid_OUT2.txt

This starts printing in my terminal window:
Analysing input arguments
/usr/bin/mafft: 1409: /usr/bin/mafft: cannot open /tmp/mafft.F6kB5aUNQh/pre: No such file

Here is the head of my out-file:
 No	Score	          Identifier	No_Hits	LengthCluster	LengthRank	GeneMerge	Duplication	MissingExtraSequences
  2	    0	geneid_v1.4_predicted_protein_tig00000009_pilon_2	      0	Not enough evidence	Not enough evidence	Not enough evidence	Not enough evidence	Not enough evidence
  4	    0	geneid_v1.4_predicted_protein_tig00000016_pilon_1	      0	Not enough evidence	Not enough evidence	Not enough evidence	Not enough evidence	Not enough evidence
  1	   60	geneid_v1.4_predicted_protein_tig00000009_pilon_1	     14	80 [192, 776]	0% (too short)	0.0	1.0	Unexpected error
  3	   60	geneid_v1.4_predicted_protein_tig00000009_pilon_3	     39	82 [777, 1416]	0% (too short)	0.0	1.0	Unexpected error
  5	   90	geneid_v1.4_predicted_protein_tig00000016_pilon_2	    142	405 [303, 667]	44%	-0.1	1.0	Unexpected error
  6	   90	geneid_v1.4_predicted_protein_tig00000017_pilon_1	    258	144 [61, 570]	8% (too short)	0.0	1.0	Unexpected error
  8	    0	geneid_v1.4_predicted_protein_tig00000017_pilon_3	      0	Not enough evidence	Not enough evidence	Not enough evidence	Not enough evidence	Not enough evidence
 10	    0	geneid_v1.4_predicted_protein_tig00000026_pilon_1	      0	Not enough evidence	Not enough evidence	Not enough evidence	Not enough evidence	Not enough evidence
  7	   90	geneid_v1.4_predicted_protein_tig00000017_pilon_2	    500	242 [56, 852]	16% (too short)	0.1	1.0	Unexpected error

**Comment by IsmailM** [comment]
Hey @nldowell , 

The following error message suggests that Mafft is not working for you...
```bash
/usr/bin/mafft: 1409: /usr/bin/mafft: cannot open /tmp/mafft.F6kB5aUNQh/pre: No such file
```

<br>
In order to help us get to the root of the issue, can you answer the following questions:

1) What version of mafft is installed? (i.e. run the following)
	
```bash
$ /usr/bin/mafft --version
```

2) Can you run mafft outside Genevalidator (i.e. run  with this [exemplar file](http://mafft.cbrc.jp/alignment/software/ex2.txt) )

```bash
$ /usr/bin/mafft ex2.txt > aligned.fa
```


3) Can you write to `/tmp/` - run the following commands
```bash
# The following should complete successfully
$ touch /tmp/exemplar.test
# The following should show you the path of the temp file
$ mktemp -dt exemplar2.test 
```

4) Do you still see the issue when using a local `$TMPDIR`?
```
$ mkdir ~/temporary
$ TMPDIR=~/temporary

$ genevalidator -d /home/snaketeamalpha/Documents/noah/RefSeqData/uniRef50/uniRef50ProtDB -n 40 \
-t ../blastUniRef50/scutB_orthoSetA.geneID_proteins.fasta.uf50hits.txt \
-b /home/snaketeamalpha/lib/ncbi-blast-2.3.0+/bin/ \
-b /usr/bin/ \
-o 'qseqid sseqid sacc slen qstart qend sstart send length qframe pident nident evalue qseq sseq' \
-r ../blastUniRef50/scutB_orthoSetA.geneID_proteins.fasta.uf50hits.txt.raw_seq \
../proteinSeqs/scutB_orthoSetA.geneID_proteins.fasta > scutB_orthoSetA_geneValid_OUT2.txt

 ```

**Comment by nldowell** [comment]
Thank you IsmailM!

We were running an old version of MAFFT and the environmental variable '$MAFFT_BINARIES' was set to an odd location.

So for anyone who stumbles here I did the following (which I should have tried first):

1. Install the latest version (v7.273) of MAFFT

2. Add the following line to my .profile file:
export MAFFT_BINARIES="/usr/libexec/mafft/"

3. source .profile

4. Then I ran mafft on your 'ex2.txt' file and it worked fine. [Note that I ran this with the old version too. And it works fine so I think the primary problem was the odd location we had the MAFFT_BINARIES in.]

5. I then removed the initial *.html file and re-ran the genevalidator command.
Original error you quoted above is fixed. Also the 'unexpected error' in the output file is no longer present.

I can check the output more extensively when it finishes.

Thanks again for all of your help.

-Noah
